### PR TITLE
feat: display AI Fix ⚡️ [IDE-580]

### PR DESCRIPTION
**Pull request by cat2608**
### Description

Wire-up CTAs (call-to-action):

- Generate fix using Snyk DeepCode AI
- Apply fix
- Retry generating AI fixes

This change needs: https://github.com/snyk/snyk-ls/pull/647

#### Tickets:
- IDE-579
- IDE-580

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+                    withContext(Dispatchers.Main) {
+                        jbCefBrowser.cefBrowser.executeJavaScript(script, jbCefBrowser.cefBrowser.url, 0)
+                    }
+                } catch (e: Exception) {
+                    e.printStackTrace()
+                }
+            }
+            return@addHandler JBCefJSQuery.Response("success")
+        }
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:70]: You can also use `runAsync`. I'm confident that `kotlinx` is available in all IDEA IDEs - but we need to test this in a non Kotlin IDE, e.g. Goland/Pycharm.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    fun generateApplyFixCommand(jbCefBrowser: JBCefBrowserBase): CefLoadHandlerAdapter {
+        val applyFixQuery = JBCefJSQuery.create(jbCefBrowser)
+
+        applyFixQuery.addHandler { value ->
+            val params = value.split(":")
+            val filePath = params[0]  // Path to the file that needs to be patched
+            val patch = params[1]      // The patch we received from LS
+
+            println("[applyFixHandler] Received request to apply fix on file: $filePath")
+            println("[applyFixHandler] Patch to apply: $patch")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:50]: this should go to a logger. Have a look at e.g. LanguageServerWrapper how to retrieve an IntelliJ logger.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    private fun applyPatchAndSave(project: Project, filePath: String, patch: String): Boolean {
+        val virtualFile = findVirtualFile(filePath) ?: run {
+            println("[applyPatchAndSave] Virtual file not found for path: $filePath")
+            return false
+        }
+
+        println("[applyPatchAndSave] Found virtual file: $virtualFile")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:95]: same here

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+        val fileContent = ApplicationManager.getApplication().runReadAction<String?> {
+            val document = FileDocumentManager.getInstance().getDocument(virtualFile)
+            document?.text
+        } ?: run {
+            println("[applyPatchAndSave] Document not found or is null for virtual file: $filePath")
+            return false
+        }
+
+        println("[applyPatchAndSave] Initial file content: $fileContent")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:105]: same ;)

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+        val diffPatch = parseDiff(patch)
+        val patchedContent = applyPatch(fileContent, diffPatch)
+
+        println("[applyPatchAndSave] Patched content that will be written: $patchedContent")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:110]: now marking every println ;)

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+        // Apply the patch inside a WriteCommandAction
+        WriteCommandAction.runWriteCommandAction(project) {
+            val document = FileDocumentManager.getInstance().getDocument(virtualFile)
+            document?.let {
+                it.setText(patchedContent)
+
+                println("[applyPatchAndSave] Content after applying patch: ${it.text}")
+
+                FileDocumentManager.getInstance().saveDocument(it)
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:120]: We should not save the document after fixing. This can cause problems for user workflows.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+        println("[applyPatchAndSave] Patched content that will be written: $patchedContent")
+
+        // Apply the patch inside a WriteCommandAction
+        WriteCommandAction.runWriteCommandAction(project) {
+            val document = FileDocumentManager.getInstance().getDocument(virtualFile)
+            document?.let {
+                it.setText(patchedContent)
+
+                println("[applyPatchAndSave] Content after applying patch: ${it.text}")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:118]: ...

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+                FileDocumentManager.getInstance().saveDocument(it)
+                println("[applyPatchAndSave] Patch applied successfully!")
+            } ?: run {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:122]: can't this just be the log command without a run?

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+        return DiffPatch(
+            originalFile = originalFile,
+            fixedFile = fixedFile,
+            hunks = hunks
+        )
+    }
+
+    private fun findVirtualFile(filePath: String): VirtualFile? {
+        return LocalFileSystem.getInstance().findFileByPath(filePath)
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:213]: I think you can just use `filePath.toVirtualFile()` here. There should be a helper function in our Utils.kt file.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        val aiFixQuery = JBCefJSQuery.create(jbCefBrowser)
+
+        aiFixQuery.addHandler { value ->
+            val params = value.split(":")
+            val folderURI = params[0]
+            val fileURI = params[1]
+            val issueID = params[2]
+
+            // Avoids blocking the UI thread
+            CoroutineScope(Dispatchers.IO).launch {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt:28]: Same here - I'd prefer `runAsync`, to keep it similar to other places in the code. But should work.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                try {
+                    val responseDiff: List<LanguageServerWrapper.Fix> =
+                        LanguageServerWrapper.getInstance().sendCodeFixDiffsCommand(folderURI, fileURI, issueID)
+
+                    val script = """
+                        window.receiveAIFixResponse(${Gson().toJson(responseDiff)});
+                    """.trimIndent()
+
+                    withContext(Dispatchers.Main) {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt:37]: Question: what does Dispatchers.Main do? Is it invoking the UI thread similar to `invokeLater`?

File: src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                        jbCefBrowser.cefBrowser.executeJavaScript(script, jbCefBrowser.cefBrowser.url, 0)
+                        JBCefJSQuery.Response("success")
+                    }
+                } catch (e: Exception) {
+                    e.printStackTrace()
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt:42]: This needs to get real exception handling.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/Utils.kt [hunk 13]
```diff
@@ -19,7 +19,7 @@ object JCEFUtils {
         val cefClient = JBCefApp.getInstance().createClient()
         cefClient.setProperty("JS_QUERY_POOL_SIZE", 1)
         val jbCefBrowser =
-            JBCefBrowserBuilder().setClient(cefClient).setEnableOpenDevToolsMenuItem(false)
+            JBCefBrowserBuilder().setClient(cefClient).setEnableOpenDevToolsMenuItem(true)
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/Utils.kt:22]: Probably something to revert once the PR is ready for merging :)

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt [hunk 14]
```diff
@@ -59,8 +61,18 @@ class SuggestionDescriptionPanelFromLS(
                 loadHandlerGenerators += {
                     openFileLoadHandlerGenerator.generate(it)
                 }
+
+                val generateAIFixHandler = GenerateAIFixHandler(snykFile.project)
+                loadHandlerGenerators += {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt:66]: interesting syntax. haven't seen that before for functions

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt [hunk 15]
```diff
@@ -176,6 +180,193 @@ class SuggestionDescriptionPanelFromLS(
             .map { allowedChars.random() }
             .joinToString("")
     }
+
+    private fun getCustomScript(): String {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt:184]: This feels like 90% of this could in a follow-up PR be pushed down to be served directly from LS.

File: src/main/resources/stylesheets/snyk_code_suggestion.scss [hunk 16]
```diff
@@ -108,3 +108,35 @@ a,
...
+button.generate-ai-fix:hover, button.sn-apply-fix:hover {
+  opacity: 0.9; /* Slightly reduces opacity on hover */
+}
+
+button.sn-apply-fix {
+  border-radius: 3px;
+  border: 1px solid var(--dark-button-border-default);
+  background: var(--dark-button-default);
+  padding: 4px 12px;
+}
```
Comment (bastiandoetsch) [inline src/main/resources/stylesheets/snyk_code_suggestion.scss:142]: I wonder if we can just push all of this down to LS, without having to use SCSS at all.

File: src/test/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandlerTest.kt [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+                challengeUtils.solveIf(challenges.premiumPaywallChallenge, () => { return true })
+                res.sendFile(path.resolve('frontend/dist/frontend/assets/private/JuiceShop_Wallpaper_1920x1080_VR.jpg'))
+              }
+            }
+        """.trimIndent()
+
+        assertEquals(expectedPatchedContent, patchedContent)
+    }
+
+    private fun applyPatch(fileContent: String, diffPatch: DiffPatch): String {
```
Comment (bastiandoetsch) [inline src/test/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandlerTest.kt:80]: Why are the following functions here? Don't we have them elsewhere? They look like they could be duplicates.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+                    val success = applyPatchAndSave(project, filePath, patch)
+
+                    val script = """
+                        window.receiveApplyFixResponse($success);
+                    """.trimIndent()
+
+                    withContext(Dispatchers.Main) {
+                        jbCefBrowser.cefBrowser.executeJavaScript(script, jbCefBrowser.cefBrowser.url, 0)
+                    }
+                } catch (e: Exception) {
```
Comment (acke) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:65]: Handle this better!

what is the errors we get, and how do we want to act on the error, do we want to return the error, do we want to update the UX to inform user?

Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:67]: Fixed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt (hunk 12 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/GenerateAIFixHandler.kt:39]: Fixed

File: src/test/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandlerTest.kt (hunk 17 above)
Comment (acke) [reply src/test/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandlerTest.kt]: Fixed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 18 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt]: Added balloon notify here.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+            Result.success(Unit)
+        } catch (e: Exception) {
+            log("[applyPatchAndSave] Error applying patch to: $filePath. e: $e")
+            Result.failure(e)
+        }
+    }
+
+    private fun log(logMessage: String) {
+        when {
+            enableDebug -> logger.debug(logMessage)
```
Comment (ShawkyZ) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:133]: why is this necessary? 
Shouldn't the logger handle loglevels ?

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt (hunk 15 above)
Comment (ShawkyZ) [reply src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt:184]: Agree
VS will need this too.
Also we will have to change syntax back to es5

File: src/main/kotlin/snyk/common/lsp/LanguageServerWrapper.kt [hunk 20]
```diff
@@ -494,6 +497,40 @@ class LanguageServerWrapper(
...
     }
 
+    data class Fix(
+        @SerializedName("fixId") val fixId: String,
+        @SerializedName("unifiedDiffsPerFile") val unifiedDiffsPerFile: Map<String, String>
+    )
+
+    @Suppress("UNCHECKED_CAST")
+    fun sendCodeFixDiffsCommand(folderURI: String, fileURI: String, issueID: String): List<Fix> {
+        if (!ensureLanguageServerInitialized()) return emptyList()
```
Comment (ShawkyZ) [inline src/main/kotlin/snyk/common/lsp/LanguageServerWrapper.kt:507]: not part of this PR.
But we need to refactor this class

File: src/main/resources/stylesheets/snyk_code_suggestion.scss (hunk 16 above)
Comment (ShawkyZ) [reply src/main/resources/stylesheets/snyk_code_suggestion.scss:142]: Actually, I think even in LS and VSC we shouldn't be using SCSS.
It's much easier to just use CSS and avoid the compilation overhead all together.
Our CSS files are also simple for this use case.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 19 above)
Comment (ShawkyZ) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:133]: am not sure if we need this
why don't we directly use `logger<ApplyFixHandler>().trace` ?

**Review by ShawkyZ — APPROVED** [review]
Looks great !

Please check my comments. 
Nothing blocking

File: src/main/kotlin/io/snyk/plugin/DiffPatcher.kt [hunk 21]
```diff
@@ -0,0 +1,94 @@
+package io.snyk.plugin
+
+import io.snyk.plugin.ui.jcef.Change
+import io.snyk.plugin.ui.jcef.DiffPatch
+import io.snyk.plugin.ui.jcef.Hunk
+
+class DiffPatcher {
+
+    fun applyPatch(fileContent: String, diffPatch: DiffPatch): String {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/DiffPatcher.kt:9]: We should do a follow-up story: move the logic to the server-side by calling a command. Flow would be: 
- applyPatch would call a language server command ("snyk.applyUnifiedDiff")
- this would lead to the language server sending a workspace edit with `workspace/applyEdit`
- the advantage is, that we would then be able to use the existing functionality in the IDEs to apply changes to the document.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+data class Hunk(
+    val startLineOriginal: Int,
+    val numLinesOriginal: Int,
+    val startLineFixed: Int,
+    val numLinesFixed: Int,
+    val changes: List<Change>
+)
+
+sealed class Change {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:34]: Why is this sealed? Is this necessary?

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+                }
+            }
+            Result.success(Unit)
+        } catch (e: Exception) {
+            log("[applyPatchAndSave] Error applying patch to: $filePath. e: $e")
+            Result.failure(e)
+        }
+    }
+
+    private fun log(logMessage: String) {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:131]: This is unnecessary - just decide what's trace and debug information.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+    data class Addition(val line: String) : Change()
+    data class Deletion(val line: String) : Change()
+    data class Context(val line: String) : Change()  // Unchanged line for context
+}
+
+class ApplyFixHandler(private val project: Project) {
+
+    private val enableDebug = Logger.getInstance("Snyk Language Server").isDebugEnabled
+    private val enableTrace = Logger.getInstance("Snyk Language Server").isTraceEnabled
+    private val logger = Logger.getInstance(this::class.java)
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:44]: Just do sth like this:
```
val logger = Logger.getInstance(this::class.java).apply {
            // tie log level to language server log level
            if (languageServerWrapper.logger.isDebugEnabled) this.setLevel(LogLevel.DEBUG)
            if (languageServerWrapper.logger.isTraceEnabled) this.setLevel(LogLevel.TRACE)
        }
```
This sets the log level of the logger, so it only writes out messages that are on this level or above. In the code, you'd then decide which information is necessary on info, debug, trace level to manage logging verbosity adequately.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+        val applyFixQuery = JBCefJSQuery.create(jbCefBrowser)
+
+        applyFixQuery.addHandler { value ->
+            val params = value.split("|@", limit = 2)
+            val filePath = params[0]  // Path to the file that needs to be patched
+            val patch = params[1]      // The patch we received from LS
+
+            // Avoid blocking the UI thread
+            runAsync {
+                //var success = true
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:57]: this can go away

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+                    } else {
+                        log("[applyPatchAndSave] Patch did not modify content: $filePath")
+                    }
+                } else {
+                    log("[applyPatchAndSave] Failed to find document for: $filePath")
+                    return@runWriteCommandAction
+                }
+            }
+            Result.success(Unit)
+        } catch (e: Exception) {
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:125]: Would it be possible to catch more specific exceptions? If yes, that would be preferable.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+        return try {
+            WriteCommandAction.runWriteCommandAction(project) {
+                val document = FileDocumentManager.getInstance().getDocument(virtualFile)
+                if (document != null) {
+                    val originalContent = document.text
+                    val patchedContent = patcher.applyPatch(originalContent, patcher.parseDiff(patch))
+                    if (originalContent != patchedContent) {
+                        document.setText(patchedContent)
+                    } else {
+                        log("[applyPatchAndSave] Patch did not modify content: $filePath")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:117]: the logs should use the log levels in the logger. E.g., this would be debug information, while a failure should probably be a warning.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt [hunk 28]
```diff
@@ -57,20 +58,21 @@ class ThemeBasedStylingGenerator {
                 httpStatusCode: Int,
             ) {
                 if (frame.isMain) {
-                    val baseColor = UIUtil.getTextFieldBackground()
-                    val (addedColor, removedColor) = getCodeDiffColors(baseColor, isHighContrast)
+                    val baseColor = UIUtil.getTextFieldBackground() //TODO Replace with JBUI.CurrentTheme colors
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:61]: Is the TODO comment still right? It looks like it's using the theme color?

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt [hunk 29]
```diff
@@ -57,20 +58,21 @@ class ThemeBasedStylingGenerator {
...
+                    val (addedColor, removedColor) = getCodeDiffColors(baseColor, isHighContrast) //TODO Replace with JBUI.CurrentTheme colors
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:62]: Same here? Is the TODO still right?

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt [hunk 30]
```diff
@@ -57,20 +58,21 @@ class ThemeBasedStylingGenerator {
...
-                    val editorColor = toCssHex(UIUtil.getTextFieldBackground())
                     val labelColor = toCssHex(JBUI.CurrentTheme.Label.foreground())
 
                     val globalScheme = EditorColorsManager.getInstance().globalScheme
-                    val tearLineColor = globalScheme.getColor(ColorKey.find("TEARLINE_COLOR")) // The closest color to target_rgb = (198, 198, 200)
-                    val tabItemHoverColor = globalScheme.getColor(ColorKey.find("INDENT_GUIDE")) // The closest color to target_rgb = RGB (235, 236, 240)
-                    val codeTagBgColor = globalScheme.getColor(EditorColors.GUTTER_BACKGROUND)  ?: globalScheme.defaultBackground
+                    val tearLineColor = globalScheme.getColor(ColorKey.find("TEARLINE_COLOR")) //TODO Replace with JBUI.CurrentTheme colors
+                    val tabItemHoverColor = globalScheme.getColor(ColorKey.find("INDENT_GUIDE")) //TODO Replace with JBUI.CurrentTheme colors
+                    val codeTagBgColor = globalScheme.getColor(EditorColors.GUTTER_BACKGROUND)  ?: globalScheme.defaultBackground //TODO Replace with JBUI.CurrentTheme colors
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:75]: same question about theme...

Comment (bastiandoetsch) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:75]: Also, why leave the TODOs in there instead of just doing it?

File: src/main/kotlin/snyk/common/lsp/LanguageServerWrapper.kt [hunk 31]
```diff
@@ -494,6 +497,40 @@ class LanguageServerWrapper(
         return this.getFeatureFlagStatus("snykCodeConsistentIgnores")
     }
 
+    data class Fix(
```
[the remark this answers is not in the document]
Comment (acke) [reply src/main/kotlin/snyk/common/lsp/LanguageServerWrapper.kt:500]: Fixed.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 24 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:44]: Fixed

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 25 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:57]: Fixed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+                    applyPatchAndSave(project, filePath, patch)
+                } catch (e: IOException) { // Catch specific file-related exceptions
+                    log("Error applying patch to file: $filePath. e:$e")
+                    Result.failure(e)
+                } catch (e: Exception) {
+                    log("Unexpected error applying patch. e:$e")
+                    Result.failure(e)
+                }
+
+                ApplicationManager.getApplication().invokeLater {
```
[the remark this answers is not in the document]
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:68]: Removed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 27 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:117]: Fixed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 23 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:131]: Removed!

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt (hunk 30 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:75]: @neil-snyk will provide the new colors from the JBUI color schemas.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt (hunk 26 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:125]: Fixed!

File: src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt (hunk 15 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/toolwindow/panels/JCEFDescriptionPanel.kt:184]: I have created jira ticket for this
https://snyksec.atlassian.net/browse/IDE-655

File: src/main/resources/stylesheets/snyk_code_suggestion.scss (hunk 16 above)
Comment (acke) [reply src/main/resources/stylesheets/snyk_code_suggestion.scss:142]: I have created jira ticket for this
https://snyksec.atlassian.net/browse/IDE-655

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt (hunk 28 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:61]: @neil-snyk will provide the new colors from the JBUI color schemas.

created ticket for that work here
https://snyksec.atlassian.net/browse/IDE-656

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt (hunk 29 above)
Comment (acke) [reply src/main/kotlin/io/snyk/plugin/ui/jcef/ThemeBasedStylingGenerator.kt:62]: @neil-snyk will provide the new colors from the JBUI color schemas.

File: src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt [hunk 33]
```diff
@@ -105,34 +86,21 @@ class ApplyFixHandler(private val project: Project) {
...
+                val patchedContent = patcher.applyPatch(originalContent, patcher.parseDiff(patch))
+                if (originalContent != patchedContent) {
+                    document.setText(patchedContent)
                 } else {
-                    log("[applyPatchAndSave] Failed to find document for: $filePath")
-                    return@runWriteCommandAction
+                    logger.warn("[applyPatchAndSave] Patch did not modify content: $filePath")
                 }
+            } else {
+                logger.error("[applyPatchAndSave] Failed to find document for: $filePath")
```
Comment (bastiandoetsch) [inline src/main/kotlin/io/snyk/plugin/ui/jcef/ApplyFixHandler.kt:100]: Shouldn't this cause a balloon notification to inform the user?
